Spatial_Engine/Query/DominantVector.cs

```diff
@@ -0,0 +1,154 @@
+        [Input("orthogonalTolerance", "Optional, when orthogonalPriority is true it will only return orthogonal vector when its length is higher this number multiplied by the actual dominant vector lengths. For example if the dominant vector is 10 in length but the orthogonal is only 5 in length, then this number should be 0.5.")]
+        [Input("angleTolerance", "Optional, angle in radians that vectors will be considered similar. Default value in radians is approximately 5 degrees.")]
+        [Output("dominantVector", "The the dominant vector of an Element1D.")]
+        public static BH.oM.Geometry.Vector DominantVector(IElement1D element1D, bool orthogonalPriority = true, double orthogonalLengthTolerance = 0.5, double angleTolerance = 0.087)
+        {
+            List<Vector> vectors = new List<Vector>();
+            List<ICurve> curves = element1D.IGeometry().ISubParts().ToList();
+
+            foreach (ICurve curve in curves)
+            {                
+                //get vector from curves and make all of them positive to better group similar vectors 
+                double x = curve.IStartDir().X;
+                double y = curve.IStartDir().Y;
+                double z = curve.IStartDir().Z;
+
+                Vector vector = BH.Engine.Geometry.Create.Vector(x, y, z);
+                vectors.Add(vector*curve.Length());
+            }
+
+            //group vectors by direction whilst comparing angle for tolerance
+            List<List<Vector>> groupByNormal = GroupSimilarVectorsWithTolerance(vectors, angleTolerance);
+
+            //then sum their total length
+            List<double> groupedLengths = groupByNormal.Select(x => x.Select(y => y.Length()).Sum()).ToList();
+
+            //get index of biggest length, which will be dominant vector
+            int biggestLengthIndex = groupedLengths.IndexOf(groupedLengths.Max());
+
+            Vector dominantVector = groupByNormal[biggestLengthIndex].First().Normalise();
+            
+            if(!orthogonalPriority)
+                return dominantVector;
+
+            if (dominantVector.X == 0 || dominantVector.X == 1)
+                return dominantVector;
+
+            //filter grouped vectors to find only curves that are X = 0 or 1 (horizontal or vertical lines)
+            var orthogonalVectors = groupByNormal.Where(x => x.First().Normalise().X == 0 || x.First().Normalise().X == 1).ToList();
+            //then sum their total length
+            List<double> orthogonalLengths = orthogonalVectors.Select(x => x.Select(y => y.Length()).Sum()).ToList();
+            //get index of biggest length, which will be dominant vector
+            int biggestOrthogonalLengthIndex = orthogonalLengths.IndexOf(orthogonalLengths.Max());
+            
+            Vector orthogonalDominantVector = orthogonalVectors[biggestOrthogonalLengthIndex].First().Normalise();
+            if (dominantVector.IsEqual(orthogonalDominantVector))
+                return dominantVector;
```

Comment:
These 2 lines seem to be redundant, the next line will result in same output anyways.